src/tracks/MidiClip.cpp

```diff
@@ -208,13 +208,13 @@ Note * MidiClip::addNote( const Note & _new_note, const bool _quant_pos )
 
 
 
-void MidiClip::removeNote( Note * _note_to_del )
+NoteVector::const_iterator MidiClip::removeNote(Note* note_to_del)
 {
 	instrumentTrack()->lock();
 
-	m_notes.erase(std::remove_if(m_notes.begin(), m_notes.end(), [&](Note* note)
+	auto it = m_notes.erase(std::remove_if(m_notes.begin(), m_notes.end(), [&](Note* note)
```

Comment:
Does the erase/remove_if idiom always return a new `end()` iterator? `std::remove_if` shifts the element(s) it removes to the end of the vector, and `erase` returns the iterator past the end which would be `end()`. This would cut iteration over the PianoRoll loop short, which might not be what we want.

If this method only ever removes 0 or 1 elements from the `m_notes` vector (and no more than that), maybe we could use `std::find_if` instead of `std::remove_if`. Then `erase` should return an iterator to the element after the note that was removed.

Reply:
Actually, there should be a version of `removeNote()` which takes an iterator rather than `Note*`. That would be more efficient since we already have the iterator we need at the call site in PianoRoll.

Reply:
> Actually, there should be a version of `removeNote()` which takes an iterator rather than `Note*`. That would be more efficient since we already have the iterator we need at the call site in PianoRoll.

Entirely possible, but we can keep that for a future refactor. For now this needs to be fixed for anyone using master builds can segfault on deleting notes, definitely not desirable. :)


> This would cut iteration over the PianoRoll loop short, which might not be what we want.

No, erase returns the next item in the vector. In the piano roll loop we don't increment the iterator in the instance we remove a note. Therefore we don't skip any notes. All good. 👍

> If this method only ever removes 0 or 1 elements from the `m_notes` vector (and no more than that), maybe we could use `std::find_if` instead of `std::remove_if`. Then `erase` should return an iterator to the element after the note that was removed.

This will delete any note at the position I believe, but I haven't checked/tested. So it could delete more than 1 note from visual reading of the loop.

Reply:
> No, erase returns the next item in the vector.

Correct. The problem is that `std::remove_if` moves all the elements that will *not* be removed to the beginning of the vector and returns an iterator to the element past those elements - the first element to be erased. Then, when that iterator is passed to `erase`, it erases all those elements at the end and returns an iterator past the end - which is the new `end()` iterator. So I don't think using it in a loop will work. It will just terminate the loop.

Reply:
Oh you're saying remove `remove_if`?

Reply:
No, I think there should be a method like this if there isn't one already:

```cpp
NoteVector::const_iterator MidiClip::removeNote(NoteVector::const_iterator iter)
{
	instrumentTrack()->lock();

	delete *iter;
	auto it = m_notes.erase(iter);

	updateLength();

	emit dataChanged();
	return it;
}
```

And then the changes you made to `removeNote()` could be reverted.

Reply:
If the current `removeNote()` only ever removes 0 or 1 elements, it *could* use `std::find_if` instead of `std::remove_if` I think, but it would be less efficient than just removing the note using the iterator which we already have.

Reply:
Sure, if we want to do that instead we can. We should still return the iterator from `removeNote(Note*)` though.

Reply:
> We should still return the iterator from removeNote(Note*) though.

There's not much point if it isn't used. And keeping the current behavior where it always returns `end()` would be worse than it returning nothing.

Reply:
The current behavior returns nothing.

Reply:
Sorry, by current behavior I meant the current state of this PR